# SkewC
Latest version of SkewC method implemented in R SkewC a novel quality-assessment method to identify poor quality single-cells in scRNA-seq experiments. 
The method relies on the measure of skewness of the gene coverage of each single cell as a quality measure. This folder contains several R Markdowns and should be run in the same order as listed below:

1-SkewC_Create_Coverage_Matrixes.Rmd
To run this, 
      a.user need to have the coverage file (coverage. r) Which contains the results of the gene body coverage. We provide example coverage. r file.
      b-User will be asked to enter the dataset title. This is will be used as label in the output plots.
      
2-SkewC_Plot_Gene_Body_Coverage.Rmd
     This script will generate two PDFs

3-SkewC_TrimClustering.Rmd 
  User will be asked to supply the Alpha value which should be a value between [0 .. 1]. E.g. for large dataset generated by 10X genomics     start with Alpha = 0.099 and evaluate the result.
  This is script will generate two text file (.tsv), contains the list of Typical and Skewed cells.
   Also it will generate a plot with the clustering result (PDF)

4-SkewC_Plot_Typical_Skewed_Coverage.Rmd
  This R Markdown will plot the gene body coverage of the Typical and Skewed cells. 
